### DISCUSSION TITLE
"unknown symbol" when trying to run/test python opened

### DISCUSSION THREAD

**Discussion by jpgoldberg** [opendp/opendp#486] [OpenDP Library] [answered]
I am trying to build opendp on macOS with an arm64 (Apple M1 Max) chip.

The rust build appears to work just fine, and everything passes with `cargo test`. This creates, among other things, `opendp/rust/target/debug/libopendp.dylib`.

When I follow the documentation instructions for building the python tools, the build process appears to go fine, but when I run pytest, I get 70 errors, all of which are of the form of "unknown symbol".

I do not understand the Rust macros well enough to really grok how the FFI is built, but I note that the generated dynamic library does not appear to have any public symbols.

```
% otool -TV rust/target/debug/libopendp.dylib
rust/target/debug/libopendp.dylib:
Table of contents (0 entries)
module name      symbol name
%
```

I similarly get an empty list with `nm -gU`. 

So I expect nothing is actually being exported from `libopendp.dylib`, but I don't know why.

**Comment by Shoeboxam** [comment, marked as answer]
This is an issue with our documentation for feature flags. 

Try instead running `cargo build --features untrusted,ffi`. The `untrusted` feature flag includes all constructors in contrib, and the `ffi` feature flag adds the external C ffi functions to the resulting .dylib. 

If you are making new FFI functions/editing bootstrap.json files, then compiling with the `bindings-python` feature will rebuild the meas.py, trans.py and comb.py files.

**Reply by Shoeboxam** [reply]
The development environment documentation on the "latest" branch is a bit more helpful:
https://docs.opendp.org/en/latest/developer/development-environment.html#building-opendp

It will become "stable" when we make our next release. Sorry for the confusion, and please feel free to more questions if you encounter any other issues!

**Reply by jpgoldberg** [reply]
Thank you very much. I really should have figured that out from `Cargo.toml`, but my initial guess was that I was running into an Apple M1 architecture problem (as I've certainly had some of those), so I went exploring down the wrong path.

Well, at least I've tested linking to Rust created dynamic libraries from C, C++, and Python on my system in my efforts to where the problem was. And as all of those worked, I was coming to the conclusion that the ffi module(?) wasn't building.

**Reply by jpgoldberg** [reply]
And just to confirm, all tests (including the python ones) now pass.

I do not know how to package things, but as I can now build a python opendp for darwin-arm64 let me know if it would be helpful for some distribution.

**Reply by Shoeboxam** [reply]
[I opened another PR because I think our docs could be clearer.](https://github.com/opendp/opendp/pull/488)

Thanks for the offer! Actually, the 0.4.0 binaries on PyPi have had issues with M1. We have updated the CI processes to build a second mac binary for M1 with cross-compilation, and we lipo the two binaries together ([relevant PR](https://github.com/opendp/opendp/pull/441)). That should come in 0.5.0. Is this why you were building from source?
